### ISSUE TITLE
Resolver: HTTPS client never closes connections

### ISSUE THREAD

**Issue by balboah** [bluejekyll/trust-dns#1043]
**Describe the bug**
When configuring using `NameServerConfigGroup::from_ips_https()` the resulting client cause connections to keep ESTABLISHED state and cause a memory leak adding a new connection for each lookup.

My loop of 200 lookups caused 200 connections in ESTABLISHED state.

**To Reproduce**

Running this example prints 196-200:

<details><summary>example main.rs</summary>

```rust
use std::net::IpAddr;
use std::process::Command;

use tokio::runtime::Builder;

use trust_dns_client::rr::{Name, RecordType};
use trust_dns_resolver::config::ResolverConfig;
use trust_dns_resolver::config::{NameServerConfigGroup, ResolverOpts};
use trust_dns_resolver::TokioAsyncResolver;

fn main() {
  let mut runtime = Builder::new()
    .threaded_scheduler()
    .core_threads(1)
    .enable_all()
    .build()
    .unwrap();

  let dns_name = "cloudflare-dns.com";
  let dns_ips: Vec<IpAddr> = vec!["1.1.1.1".parse().unwrap()];
  let ns = NameServerConfigGroup::from_ips_https(&dns_ips, 443, dns_name.to_string());
  let options = ResolverOpts {
    preserve_intermediates: true,
    ..ResolverOpts::default()
  };
  let config = ResolverConfig::from_parts(None, vec![], ns);
  let resolver = runtime
    .block_on(TokioAsyncResolver::new(
      config,
      options,
      runtime.handle().clone(),
    ))
    .unwrap();

  for n in 0..200 {
    let name: Name = format!("{}.zendesk.com.", n).parse().unwrap();
    runtime
      .block_on(resolver.lookup(name, RecordType::A, Default::default()))
      .unwrap();
  }

  let mut netstat = Command::new("netstat");
  let output = netstat
    .arg("-na")
    .output()
    .expect("failed to execute process");
  let output = String::from_utf8_lossy(&output.stdout);

  let connections: Vec<_> = output
    .lines()
    .filter(|l| l.contains("ESTA"))
    .filter(|l| l.contains("1.1.1.1"))
    .collect();
  println!("{}", connections.len());
}
```

</details>

**Expected behavior**

Connections should be closed after resolved lookup, which works with the TLS a.k.a DoT version.

**System:**
 - OS: darwin
 - Architecture: x86_64
 - rustc version: 1.41.1

**Version:**
Version: trust-dns git rev 77fd933d

**Additional context**
Using the rustls TLS only edition works fine.
Tokio runtime is created with:

```rust
Builder::new()
    .threaded_scheduler()
    .core_threads(1)
    .enable_all()
    .build()
```

**Comment by bluejekyll** [comment]
> resulting client cause connections to keep ESTABLISHED state and cause a memory leak

This is surprising. The design isn't to drop the connections between usage, in fact TCP and TLS should maintain connections, so the behavior there seems unexpected there as well. I think there are tests that cover this, but we should do a better review here.

The bigger concern is that you're seeing a memory leak, and we should determine why that is.

**Comment by balboah** [comment]
yes the optimal for my HTTPS case would be to have one keep alive connection, especially if it's able to multiplex queries on the h2 streams. I'm not sure how that works with TLS only.

but the critical part is that it leaks and triggers a crash in my limited environment

**Comment by balboah** [comment]
I haven't dug into the code yet but I remember there is a "max idle connections per host" setting in reqwest which defaults to max int size. Maybe this is a similar issue, rapid queries isn't limited to number of idle connections

**Comment by balboah** [comment]
Currently suspecting that the bug is around the `NameServerPool::parallel_conn_loop`.
There will be a new `HttpsClientStream` for every query that was not cached, it wouldn't be possible for it to re-use its connection then. I'm having trouble following the Futures calls and all the different types that collectively gets the job done so I'll leave this for now

**Comment by bluejekyll** [comment]
Yes. Some of these are older futures. All the code hasn’t been 100% cleaned up, so I can understand it being hard to follow.

I’ll take a look at that. I spent a lot of time last release cleaning up that area, so It’s fresh in my mind.

**Comment by balboah** [comment]
@bluejekyll hey, did you find some time to squash this bug? No stress, but would save me a lot of time :)

**Comment by bluejekyll** [comment]
Not yet. I was hoping to find some time this week.

**Comment by bluejekyll** [comment]
I've been looking at this today. I think what happened is that through all the refactoring between 0.18 and 0.19 to support async/await, this area of the code was screwed up. I have a test I'm building to try and detect this so that we don't lose this functionality in the future. Then I'll work on a patch to fix.
